shell/test.sh

```diff
@@ -1,114 +1,193 @@
+
+# COVER_FLAGS is an array of flags to pass to `go test -cover`.
+# COVER_FLAGS must be a string value. It will be split on spaces.
+#
+# Why: We need to support importing flags from a string variable, e.g.,
+# CircleCI in-line environment variables or contexts.
+# shellcheck disable=SC2206
+COVER_FLAGS=(${COVER_FLAGS:-})
+
+# WITH_COVERAGE is a boolean flag that enables coverage reporting. If
+# set at all, a coverage file will be generated during the test. It
+# will be outputted to /tmp/coverage.out. When running in CI, this is
+# always enabled.
+WITH_COVERAGE="${WITH_COVERAGE:-}"
+
+# GO_TEST_TIMEOUT is the timeout to pass to `go test -timeout`. If not
+# set, no timeout will be passed.
+GO_TEST_TIMEOUT="${GO_TEST_TIMEOUT:-}"
+
+# RACE determines if 'go test -race' should be enabled or not. If set to
+# 'disabled', race condition testing will not be enabled and the flag
+# will not be passed to 'go test'. Defaults to 'enabled'.
+RACE="${RACE:-enabled}"
+
+if [[ -n $CI ]]; then
+  GOFLAGS+=(-mod=readonly)
+  WITH_COVERAGE="true"
+
+  # Ensure that all processes recieve the value of GOFLAGS.
+  export GOFLAGS
+fi
+
+# If GO_TEST_TIMEOUT is set, we pass it to `go test` as a timeout.
+if [[ -n $GO_TEST_TIMEOUT ]]; then
+  TEST_FLAGS+=(-timeout "$GO_TEST_TIMEOUT")
+fi
 
 # shellcheck source=./lib/logging.sh
 source "$DIR/lib/logging.sh"
 
 # shellcheck source=./lib/bootstrap.sh
 source "$DIR/lib/bootstrap.sh"
 
-# This comes from the Makefile. This makes the linter aware of it.
-export TEST_TAGS=$TEST_TAGS
-
-if [[ -z $TEST_PACKAGES ]]; then
-  TEST_PACKAGES="./..."
-fi
-
-if [[ -n $CI ]]; then
-  export GOFLAGS="${GOFLAGS} -mod=readonly"
-fi
-
-if [[ -n $GO_TEST_TIMEOUT ]]; then
-  export TEST_FLAGS=${TEST_FLAGS:--timeout "$GO_TEST_TIMEOUT"}
-fi
+# REPODIR is the base directory of the repository.
+REPODIR=$(get_repo_directory)
 
 # Catches test dependencies by shuffling tests if the installed Go version supports it
 currentver="$(go version | awk '{ print $3 }' | sed 's|go||')"
 requiredver="1.17.0"
 if [[ "$(printf '%s\n' "$requiredver" "$currentver" | sort -V | head -n1)" == "$requiredver" ]]; then
-  TEST_FLAGS="$TEST_FLAGS -shuffle=on"
+  TEST_FLAGS+=(-shuffle=on)
 fi
 
 # Although we highly encourage all projects to run test with a check for race coditions, teams can
 # choose to temporarily disable this option
 if [[ $RACE != "disabled" ]]; then
-  TEST_FLAGS="$TEST_FLAGS -race"
+  TEST_FLAGS+=(-race)
 fi
 
-if [[ -n $WITH_COVERAGE || -n $CI ]]; then
-  COVER_FLAGS=${COVER_FLAGS:- -coverpkg=./... -covermode=atomic -coverprofile=/tmp/coverage.out -cover}
+# If WITH_COVERAGE is set, we pass flags to 'go test' to enable coverage
+# file generation.
+if [[ -n $WITH_COVERAGE ]]; then
+  COVER_FLAGS+=(-coverpkg=./... -covermode=atomic -coverprofile=/tmp/coverage.out -cover)
 fi
 
 testInclude="$(get_repo_directory)/scripts/test.include.sh"
 if [[ -e $testInclude ]]; then
+  info "Running project specific test file: $testInclude"
```

Comment:
This should help make it more clear to a user when a project-specific test file is failing vs the devbase framework.